Add test framework specification (#565)
ballerina-platform/ballerina-spec: test/test-framework-spec.md

# Ballerina Test Framework Specification

**Contributors** :

> Joseph Fonseka, Asma Jabir, Aquib Zulfikar, Fathima Dilhasha

Reviewer: @hevayo Shall we add initial contributions as well?

Reviewer 2: Yeh +1

Author: Removed the contirbutors to be consistent with other spec documents

<table>
  <tr>
   <td>@test:assertEquals
   </td>
   <td>Checks if the specified value is equal to the expected value
   </td>
  </tr>
  <tr>
   <td>@test:assertNotEquals
   </td>
   <td>Checks if the specified value is not equal to the expected value
   </td>
  </tr>
  <tr>
   <td>@test:assertTrue
   </td>
   <td>Checks if the specified value is true
   </td>
  </tr>
  <tr>
   <td>@test:assertFalse
   </td>
   <td>Checks if the specified value is false
   </td>
  </tr>
  <tr>
   <td>@test:assertFail
   </td>
   <td>Forces a test case to fail
   </td>
  </tr>
</table>

Each assertion allows providing an optional assertion fail message. 

**Example**

```ballerina
@test:Config {}
function testAssertIntEquals() {
    int answer = 0;
    int a = 5;
    int b = 3;
    answer = intAdd(a, b);
    test:assertEquals(answer, 8, msg = "int values not equal");
}
```

### Grouping Tests

Organizing tests into groups allows the user to run a specific group of tests using the `--groups` option. 

**Example**

```ballerina
@test:Config { groups: ['group1'] }
function test1() {
    // Test code ...
}
```

### Enabling / Disabling Tests

> Individual test cases can be disabled by adding the `disabled` option in the test config annotation. This makes sure that the particular test case is skipped when running the tests. 

Reviewer 3: instead of 'disabled', let's add as 'enable:false' option

```ballerina
@test:Config { enable : false }
function test1() {
    // Test code ...
}
```

When a test case is disabled, before and after functions specified in the test configurations and all the dependent
 test cases will be skipped. 

### Order of Execution

> The depends on attribute can be used to define a list of function names that the test function depends on. These functions will be executed before the test execution. The order in which the comma-separated list appears has no prominence and thus will be executed in an arbitrary manner. This attribute can be used to ensure that the tests are being executed in the expected order.

Reviewer 3: Shall we change the order of last 2 sentences? Instead of "the comma separated list", shall we use "comma separated list of functions" ?

```ballerina
@test:Config { }

> function test2() {

Reviewer 4: I think this should be test1()

      --- main.bal       
      --- resources/     
      --- tests/           <- tests for this module (e.g. unit tests)
        ---- main_test.bal <- test file for main
        ---- resources/    <- resources for these tests
```

The Ballerina test framework will only execute tests defined inside the `tests/` directory of a module. Tests defined
 outside the test directory will not get executed when building the Ballerina project. Test files can be put into subdirectories within the tests folder much like a Ballerina module.

### Visibility

The symbols defined in a module are accessible from within the test files. This includes globally defined objects and variables. Hence, redefining a symbol in a test file is not allowed if it is already declared in the module. Instead they can be reassigned in test files. It must be noted that symbols defined in the test files will not be visible inside the module source files. 

### Test Resources

The test resources folder is meant to contain any files or resources that are required for testing. From the test
 code, test resources can be accessed using the absolute path or the path relative to the project root. 

## Writing Test Cases

> A test case is a Ballerina function preceded by a test annotation which is provided by the _ballerina/test_ module. The

Author: Fixed in 8708dfa

 along with their attributes.
 
<table>
  <tr>
   <td>@test:BeforeSuite {}
   </td>
   <td>Function specified will be run once before any of the tests in the test suite is run
   </td>
  </tr>
  <tr>
   <td>@test:BeforeEach {}
   </td>
   <td>Function specified will be run before every test when the test suite is run
   </td>
  </tr>
  <tr>
   <td>@test:Config {}
   </td>
   <td>Function specified is considered as a test function
<p>
Annotation value fields :
<ul>

<li>enable : { true | false } 
<ul>
 
<li>Enable or disable test
 
<li>Default : true
</li> 
</ul>

<li>before : “&lt;function name&gt;” 
<ul>
 
<li>Name of function to run before the test is run
</li> 
</ul>

<li>after : “&lt;function name&gt;” 
<ul>
 
<li>Name of the function to run after the test is run
</li> 
</ul>

<li>dependsOn : [“&lt;function name&gt;:”, … ] 
<ul>
 
<li>A list of function names on which the function depends on. These will be run before the test.
</li> 
</ul>

<li>dataProvider : “&lt;function name&gt;” 
<ul>
 
<li>Specifies the name of the function that will be used to provide the value sets to execute the test against
</li> 
</ul>

<li>groups : [“&lt;test group name&gt;”, ...] 
<ul>
 

> <li>List of group names, one or more, that this test belongs to

Reviewer 5: ```suggestion
<li>List of group names one or more that this test belongs to
```

</li> 
</ul>
</li> 
</ul>
   </td>
  </tr>
  <tr>
   <td>@test:AfterSuite {}
   </td>

>    <td>The function specified in the following annotation will be run once after all the tests in the test suite is run.

Reviewer 5: ```suggestion
   <td>The function specified in the following annotation will be run once after all the tests in the test suite are run.
```

   </td>
  </tr>
</table>

### Assertions

> The Ballerina test framework has built-in assertions that enable users to verify an actual  output against an expected output. 

Reviewer 5: ```suggestion
The Ballerina test framework has built-in assertions, which enable users to verify an actual output against an expected output. 
```

The following are the list of available assertions available in the test framework.

<table>
  <tr>
   <td>@test:assertEquals
   </td>

>    <td>Checks if the specified value is equal to the expected value

Reviewer 5: ```suggestion
   <td>Checks if the specified value is equal to the expected value.
```

   </td>
  </tr>
  <tr>
   <td>@test:assertNotEquals
   </td>

>    <td>Checks if the specified value is not equal to the expected value

Reviewer 5: ```suggestion
   <td>Checks if the specified value is not equal to the expected value.
```

   </td>
  </tr>
  <tr>
   <td>@test:assertTrue
   </td>

>    <td>Checks if the specified value is true

Reviewer 5: ```suggestion
   <td>Checks if the specified value is true.
```

   </td>
  </tr>
  <tr>
   <td>@test:assertFalse
   </td>

>    <td>Checks if the specified value is false

Reviewer 5: ```suggestion
   <td>Checks if the specified value is false.
```

   </td>
  </tr>
  <tr>
   <td>@test:assertFail
   </td>

>    <td>Forces a test case to fail

Reviewer 5: ```suggestion
   <td>Forces a test case to fail.
```

   </td>
  </tr>
</table>

Each assertion allows providing an optional assertion fail message. 

**Example**

```ballerina
@test:Config {}
function testAssertIntEquals() {
    int answer = 0;
    int a = 5;
    int b = 3;
    answer = intAdd(a, b);
    test:assertEquals(answer, 8, msg = "int values not equal");
}
```

### Grouping Tests

Organizing tests into groups allows the user to run a specific group of tests using the `--groups` option. 

**Example**

```ballerina
@test:Config { groups: ['group1'] }
function test1() {
    // Test code ...
}
```

### Enabling / Disabling Tests

Individual test cases can be disabled by adding the `enable : false` option in the test config annotation. This makes sure that the particular test case is skipped when running the tests. 

```ballerina
@test:Config { enable : false }
function test1() {
    // Test code ...
}
```

When a test case is disabled, before and after functions specified in the test configurations and all the dependent
 test cases will be skipped. 

### Order of Execution

> The depends on attribute allows the user to define a list of function names that the test function depends on. These

Reviewer 5: ```suggestion
The `depends on` attribute allows the user to define a list of function names that the test function depends on. These
```

> functions will be executed before the test execution. The order in which the comma separated list of functions appears

Reviewer 5: ```suggestion
functions will be executed before the test execution. The order in which the comma-separated list of functions appears
```

has no prominence and thus will be executed in an arbitrary manner. This attribute can be used to ensure that the
tests are being executed in the expected order.

```ballerina
@test:Config { }
function test1() {
    // Test code ...
}

@test:Config { dependsOn: ["test1"] }
function test2() {
    // Test code ...
}
```

## Mocking

The mocking support in the Ballerina test framework provides capabilities to mock a function or an object for unit testing. 
The mocking feature can be used to control the behavior of functions and objects by defining return values or
 replacing the entire object/function with a user-defined equivalent. This feature will help the user to test the
  Ballerina code independently from other modules and external endpoints.

### Mock Annotation

Initializing a function mock needs a preceding annotation in order to identify and replace the occurrence of the original function during compilation.

> This annotation is only required when mocking functions since a part of function mocking is handled in compile time. Mocking an object is completely handled in the runtime and therefore, this annotation is not required when initializing a mock for an object.

Reviewer 5: ```suggestion
This annotation is only required when mocking functions since a part of function mocking is handled during the compile time. Mocking an object is completely handled in the runtime, and thereby, this annotation is not required when initializing a mock for an object.
```

<table>
  <tr>
   <td>@test:Mock {}
   </td>
   <td>The function specified will be considered as a mock function that gets triggered every time the original function is called. 
<p>
Annotation value fields :
<ul>

<li>moduleName 
<ul>
 
<li>Default : Uses the current module
 

> <li>Name of the module where the function to be mocked resides in

Reviewer 5: ```suggestion
<li>Name of the module where the function to be mocked resides in.
```

</li> 
</ul>

<li>functionName 
<ul>
 
<li>Name of the function to be mocked
</li> 
</ul>
</li> 
</ul>
   </td>
  </tr>
</table>

### Initialization

*   Mock object 

    ```ballerina
    http:Client mockClient = <http:Client> test:mock(http:Client, mockObj = new);
    ```

*   Mock function 

    ```ballerina
    @test:Mock {
        moduleName : "ballerina/io"
        functionName : "println"
    }
    test:MockFunction mockFunc1  = new();
    ```

### Features

**Default behavior**

If a mock object or function is used without registering any cases, the default behavior would be to throw a runtime exception. 

Using the available features, the user can stub with preferred behaviors for function calls and values for member
 variables (of objects) before testing the required function. 

#### Case A: Features available in object mocking

**Basic Cases**

> 1. Provide a replacement mock object defined by the user at initialization

Reviewer 5: ```suggestion
1. Provide a replacement mock object defined by the user at initialization.
```

    ```ballerina
    http:Client mockClient = <http:Client> test:mock (http:Client, mockClient);
    ```

> 2. If function doesn't have a return type or has an optional return type then do nothing 

Reviewer 5: ```suggestion
2. If the function doesn't have a return type or has an optional return type, then do nothing. 
```

    
    ```ballerina
    test:prepare(mockClient).when("functionName").doNothing();
    ```

> 3. Provide a return value

Reviewer 5: ```suggestion
3. Provide a return value.
```

    ```ballerina
    test:prepare(mockClient).when("functionName").thenReturn(5);
    ```

> 4. Provide return value based on input

Reviewer 5: ```suggestion
4. Provide a return value based on the input.
```

    ```ballerina
    test:prepare(mockClient).when("functionName").withArguments(anydata...).thenReturn(5);
    ```
   

> 5. Mock member variables of an object

Reviewer 5: ```suggestion
5. Mock the member variables of an object.
```

    ```ballerina
    test:prepare(mockClient).getMember("member").thenReturn(5);
    ```

#### Case B : Features available in function mocking 

> 1. Provide a replacement function body

Reviewer 5: ```suggestion
1. Provide a replacement function body.
```

    ```ballerina
    test:when(mockFunc1).call("mockFuncName");
    ```

> 2. If function doesn't have a return type do nothing 

Reviewer 5: ```suggestion
2. If the function doesn't have a return type, do nothing. 
```

    ```ballerina
    test:when(mockFunc1).doNothing();
    ```

> 3. Provide a return value

Reviewer 5: ```suggestion
3. Provide a return value.
```

    ```ballerina
    test:when(mockFunc1).thenReturn(5);
    ```

> 4. Provide return value based on input

Reviewer 5: ```suggestion
4. Provide a return value based on the input.
```

    ```ballerina
    test:when(mockFunc1).withArguments(any...).thenReturn(5);
    ```

> 5. If mocking should not take place, call the real function

Reviewer 5: ```suggestion
5. If mocking should not take place, call the real function.
```

    ```ballerina
    test:when(mockFunc1).callRealFunction();

    ```

**Advance Cases**

> 6. Provide generalized inputs to accept any value for certain arguments

Reviewer 5: ```suggestion
6. Provide generalized inputs to accept any value for certain arguments.
```

    
   ```ballerina
   test:prepare(mockClient).when("functionName").withArguments("/pets", test:ANY, ...).thenReturn(5);
   ```
   ```ballerina
   test:when(mockFunc1).withArguments(test:ANY,...).thenReturn(5);
    ```

7. Provide multiple return values to be returned sequentially for each function call

    ```ballerina
    test:when(mockFunc1).thenReturnSequence(5,6,0)
    ```

#### Errors

> The cases can throw errors at the runtime for following reasons:

Reviewer 5: ```suggestion
The cases can throw errors at the runtime for the following reasons:
```

> *   All Cases - If the function is not available in the mocked type

Reviewer 5: ```suggestion
*   All Cases - If the function is not available in the mocked type:
```

*   Case A1

>     *   If function signatures are not equal

Reviewer 5: ```suggestion
    *   If the function signatures are not equal
```

>     *   If corresponding functions are not found

Reviewer 5: ```suggestion
    *   If the corresponding functions are not found
```

*   Case A2, Case B2

>     *   If the function has a return type specified

Reviewer 5: ```suggestion
    *   If the function has a return type specified in it
```

*   Case A3, Case B3

>     *   If the the return value does not match the function return type

Reviewer 5: ```suggestion
    *   If the return value does not match the function return type
```

*   Case A4, Case B4

>     *   If the number/type of arguments provided does not match the function return type

Reviewer 5: ```suggestion
    *   If the number/type of arguments provided do not match the function return type
```

    *   If the the return value does not match the function signature
*   Case A5

>     *   If the object does not have a member variable of specified name

Reviewer 5: ```suggestion
    *   If the object does not have a member variable of the specified name
```

    *   If the variable type does not match the return value
*   Case B1

>     *   If function signatures are not equal

Reviewer 5: ```suggestion
    *   If the function signatures are not equal
```

    *   If the replacing mock function is not found

#### Examples

**Case A**

> The mocking examples are written to mock the http calls of the following _main.bal_ file.

Reviewer 5: ```suggestion
The mocking examples are written to mock the HTTP calls of the following `_main.bal_` file.
```

Author: Fixed in 8708dfa
